src/transforms/reduce/transform.rs

```diff
@@ -39,13 +53,47 @@ impl ReduceState {
         }
     }
 
-    fn add_event(&mut self, e: LogEvent, strategies: &IndexMap<KeyString, MergeStrategy>) {
+    fn add_event(&mut self, e: LogEvent, strategies: &IndexMap<OwnedTargetPath, MergeStrategy>) {
         self.metadata.merge(e.metadata().clone());
 
+        for (path, strategy) in strategies {
+            if let Some(value) = e.get(path) {
+                match self.fields.entry(path.clone()) {
+                    Entry::Vacant(entry) => match get_value_merger(value.clone(), strategy) {
+                        Ok(m) => {
+                            entry.insert(m);
+                        }
+                        Err(error) => {
+                            warn!(message = "Failed to create value merger.", %error, %path);
+                        }
+                    },
+                    Entry::Occupied(mut entry) => {
+                        if let Err(error) = entry.get_mut().add(value.clone()) {
+                            warn!(message = "Failed to merge value.", %error);
+                        }
+                    }
+                }
+            }
+        }
+
         if let Some(fields_iter) = e.all_event_fields_skip_array_elements() {
-            for (path, value) in fields_iter {
-                let maybe_strategy = strategies.get(&path);
-                match self.fields.entry(path) {
+            for (mut path, value) in fields_iter {
+                // TODO: Addressed in issue 21077.
+                if path.contains("\\.") {
```

Comment:
I think this is OK since it is the only thing we were handling before this change, but technically I think the path part needs to be quoted if there are any special characters (like `-` or `*` or ` `).